packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+            index: nodeIndex,
+            useCustomElementInstance:
+              parent && isElement(parent) && parent.isDefinedCustomElement,
+          });
+          // There will be two comments per part (open+close) in the rendered
+          // output and on the client, so increment again for that
+          // nodeIndex++;
+        }
+      } else if (isElement(node)) {
+        // Whether to flush the start tag. This is neccessary if we're changing
+        // any of the attributes in the tag, so it's true for custom-elements
+        // which might reflect their own state, or any element with a binding.
+        let writeTag = false;
+        let boundAttrsCount = 0;
+
+        const tagName = node.tagName;
+        let ctor;
+
+        if (tagName.indexOf('-') !== -1) {
+          // Looking up the constructor here means that custom elements must be
+          // registered before rendering the first template that contains them.
+          ctor = customElements.get(tagName);
+          if (ctor !== undefined) {
+            // Write the start tag
+            writeTag = true;
+            // Mark that this is a custom element
+            node.isDefinedCustomElement = true;
+            ops.push({
+              type: 'custom-element-open',
+              tagName,
+              ctor,
+              staticAttributes: new Map(
+                node.attrs
+                  .filter((attr) => !attr.name.endsWith(boundAttributeSuffix))
+                  .map((attr) => [attr.name, attr.value])
+              ),
+            });
+          }
+        }
+        if (node.attrs.length > 0) {
+          for (const attr of node.attrs) {
+            if (attr.name.endsWith(boundAttributeSuffix)) {
+              writeTag = true;
+              boundAttrsCount += 1;
+              // Note that although we emit a lit-bindings comment marker for any
+              // nodes with bindings, we don't account for it in the nodeIndex because
+              // that will not be injected into the client template
+              const name = attr.name.substring(
+                0,
+                attr.name.length - boundAttributeSuffix.length
+              );
+              const strings = attr.value.split(marker);
+              const attrSourceLocation = node.sourceCodeLocation!.attrs[
+                attr.name
+              ];
+              const attrNameStartOffset = attrSourceLocation.startOffset;
+              const attrEndOffset = attrSourceLocation.endOffset;
+              flushTo(attrNameStartOffset);
+              ops.push({
+                type: 'attribute-part',
+                index: nodeIndex,
+                name,
+                strings,
+                tagName,
+                useCustomElementInstance: ctor !== undefined,
+              });
+              skipTo(attrEndOffset);
+            } else if (node.isDefinedCustomElement) {
+              // We will wait until after connectedCallback() and render all
```

Comment:
Why wait until `connectedCallback`?

Reply:
I'll improve the comments here.

We skip over static attributes for custom elements (don't include them as strings in 'text' opcodes), and instead capture the static values from the AST along with the `custom-elements-open` opcode (see above). As part of `custom-elements-open`, we call setAttribute of all static attributes into the custom element instance. Then as part of `custom-elements-attributes`, we call `renderer.connectedCallback()` (which is where we run the reflection code), and then serialize all the attributes (set either as statics or reflected by the renderer) back out to the DOM.

So the basic opcode flow when encountering CE's is:
* `text`
  * Emit open tag `<x-foo`
* `custom-element-open`
  * Create the CE instance+renderer and put on stack
  * Call `renderer.setAttribute()` for any `staticAttributes`
* `attribute-part`(s)
  * Call `renderer.setAttribute()` or `renderer.setProperty()` for `AttributePart`/`PropertyPart`s
* `custom-element-attributes`
  * Call `renderer.connectedCallback()`
  * Emit `renderer.renderAttributes()`
* `text`
  * Emit end of of open tag `>`
  * Emit `<!--lit-bindings ${nodeIndex}-->` marker if there were attribute parts
* `custom-element-children`
  * Emit `renderer.renderChildren()`
... (emit static `text` or recurse into more children)
* `custom-element-close`
  * Pop the CE instance+renderer off the stack

(And in the process of writing this up I realized we weren't actually emitting `staticAttributes`, so I added that and a test for it)